### PR TITLE
Upgrade bootstrap compiler to 1.12

### PR DESCRIPTION
**Pull request by brson**
r? @alexcrichton

### CODE REVIEW

**Comment by brson** [comment]
Wait, I don't think i did the hash right...

**Comment by brson** [comment]
`echo -n 1.12.0-beta.1 | md5sum | head -c 8`?

**Comment by brson** [comment]
ok, now it's ready.

**Comment by brson** [comment]
Aparently the version of cargo here needs to be bumped, along with cargo-revs.txt in rust-packaging.

**Comment by alexcrichton** [comment]
@bors: r+ 

This probably won't land though until https://github.com/rust-lang/cargo/issues/3005 is fixed, so we may want to avoid updating Cargo just yet

**Comment by eddyb** [comment]
This [doesn't compile AFAICT](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1514/steps/compile/logs/stdio).

@bors r-

**Comment by nagisa** [comment]
You need to clean-up the cfg(stage0) stuff in this PR as well.

**Comment by brson** [comment]
Removed the stage0 stuff. Seems to build. Still testing.

File: src/libcore/cmp.rs [hunk 1]
```diff
@@ -699,39 +699,6 @@ mod impls {
...
-            impl Ord for ! {
-                fn cmp(&self, _: &!) -> Ordering {
-                    *self
-                }
-            }
-        }
-    }
-
-    #[cfg(not(stage0))]
-    not_stage0!();
```
Comment (eddyb) [inline src/libcore/cmp.rs]: The contents of this macro need to stay around (without any `#[cfg]`).

File: src/libcore/fmt/mod.rs [hunk 2]
```diff
@@ -1363,29 +1363,6 @@ macro_rules! fmt_refs {
...
-        impl Display for ! {
-            fn fmt(&self, _: &mut Formatter) -> Result {
-                *self
-            }
-        }
-    }
-}
-
-#[cfg(not(stage0))]
-not_stage0!();
```
Comment (eddyb) [inline src/libcore/fmt/mod.rs]: Same here.

File: src/libstd/lib.rs [hunk 3]
```diff
@@ -280,10 +280,6 @@
 #![feature(zero_one)]
 #![cfg_attr(test, feature(update_panic_count))]
 
-// Issue# 30592: Systematically use alloc_system during stage0 since jemalloc
-// might be unavailable or disabled
-#![cfg_attr(stage0, feature(alloc_system))]
```
Comment (eddyb) [inline src/libstd/lib.rs]: Has the situation this handles changed?

Comment (alexcrichton) [inline src/libstd/lib.rs]: Ah yeah these need to stick around I believe

**Comment by brson** [comment]
Updated. Thanks for the review @eddyb .

**Pull request merged**